docs/IPC.md

```diff
@@ -18,20 +24,296 @@ Note that "persistence" here means surviving the loss of a _process_, not a node
+
+Each change in the table is expected to be updated via an update in the channel.
+
+The `DriverSubscriber` is expected to retrieve the current state of the table, and update the calling
+`pubsub.Subscription` of all changes by sending updates on a channel passed during the `DriverSubscriber`
+initialization.
+
+## Driver Implementations
+
+eve-os currently has one primary driver [socketdriver](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub/socketdriver).
+In addition, an [emptydriver](https://github.com/lf-edge/eve/blob/8c6d4ddecf5fec004d4e188f9abc03644c2746aa/pkg/pillar/pubsub/emptydriver.go)
+provides a zero-functionality implementation, which is useful for working with services that require a pubsub but will not be exercising it at all.
+
+Additional implementations may exist in testing, e.g. in-memory drivers.
+### `socketdriver`
+
+The primary `Driver` implemented in eve-os is the [socketdriver](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub/socketdriver), which uses:
+
+* data storage via files, one directory per table, one file per key-value entry; it does not use in-memory storage.
+* notification via Unix-domain sockets, one socket per table.
+
+#### Data Storage
+
+The socketdriver stores all data for one table in a specific directory. It does not use in-memory storage.
+This file mechanism is used both for persistent and for non-persistent data. For persistent data,
+it stores files in a directory that will be persisted beyond reboot, whereas for non-persistent data, it stores
+files in an ephemeral directory that will be deleted on reboot.
+
+The root directory upon which socketdriver operates is passed to the socketdriver
+instance upon creation. In the case of normal eve-os operation, that defaults to `/`, but it can be changed upon initialization.
+All other directories are subsidiary to the root directory.
+
+The specific subdirectory used inside the root directory is determined based on the options [here](https://github.com/lf-edge/eve/blob/6160d0e96c72a1954db2a8bdfd99c2fec1972341/pkg/pillar/pubsub/socketdriver/driver.go#L88-L99).
+
+The directory to use and the type of file depends on 2 options:
+
+* `persistent`: whether or not this table should persist, and therefore if files should be in a persisted-beyond-reboot directory or not.
+* `publishToDir`: whether or not this table should publish its data to a directory.
+
+It also includes the "name" of the publication, where `<name>` is calculated:
+
+* global table: `global`
+* `agentScope == ""`: `<agentName>/<topic>`
+* otherwise: `<agentName>/<agentScope>/<topic>`
+
+The above combine to create the location:
+
+1. persistent AND publish: `/persist/config/<name>`
+1. persistent AND NOT publish: `/persist/<name>`
+1. NOT persistent AND publish: `/persist/global/<name>`
```

Comment:
Small corrections:
- persistent AND NOT publish: `/persist/status/<name>`
- NOT persistent AND publish: `/run/global/<name>`"

But `publishToDir` is very confusing. After all, it is always written to some directory.
`publishToDir` is determined by `global`, so maybe it could be explained by "persistent AND NOT global", etc.
Also there is also a hidden assumption that for persistent data:
 - `global` is used for configuration
 - other key names are for state data

At least that's what the code is trying to tell. I don't know, it is confusing.

Reply:
> If it is more clear to rename/restructure things in the code to distinguish between those similar but different uses, then please go ahead.

It probably would be clear to rename, but I am trying solely to tackle understanding the current state right now. So let's not hit that quite yet.

Reply:
@eriknordmark I am going to try and take that bullet list of yours apart, to see if I get it.

### persist vs non-persist

It sounds like you are saying, for non-persistent, subscribers uses the socket to listen to publications directly from the publisher which sends packets to the socket, whereas for persistent, publishers write to the file and subscribers look for changes in the file.

But I thought that wasn't the case; I thought everything used the socket, and it just was a question of what gets done in addition to the socket.

> publisher does not need to be present when the subscriber starts

I didn't think about that. Does it then use the socket to get future updates?

> or all publications we checkpoint the data to a file (in /run). This is also used to populate the publication on start

If the pub starts after the sub, then the file won't be populated (as `/run` is ephemeral), so what does it buy us?